rotkehlchen/chain/arbitrum_one/modules/umami/decoder.py

```diff
@@ -0,0 +1,175 @@
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.arbitrum_one.node_inquirer import ArbitrumOneInquirer
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+DEPOSIT_TOPIC: Final = b'\xdc\xbc\x1c\x05$\x0f1\xff:\xd0g\xef\x1e\xe3\\\xe4\x99wbu.:\tR\x84uED\xf4\xc7\t\xd7'  # noqa: E501
+TRANSFER_TOPIC: Final = b'\xdd\xf2R\xad\x1b\xe2\xc8\x9bi\xc2\xb0h\xfc7\x8d\xaa\x95+\xa7\xf1c\xc4\xa1\x16(\xf5ZM\xf5#\xb3\xef'  # noqa: E501
+WITHDRAW_TOPIC: Final = b'\xfb\xdey} \x1ch\x1b\x91\x05e)\x11\x9e\x0b\x02@|{\xb9jJ,u\xc0\x1f\xc9fr2\xc8\xdb'  # noqa: E501
+STAKE_TOPIC: Final = b'\x90\x89\x08\t\xc6T\xf1\x1dnr\xa2\x8f\xa6\x01Iw\n\r\x11\xecl\x921\x9dl\xeb+\xb0\xa4\xea\x1a\x15'  # noqa: E501
+UNSTAKE_TOPIC: Final = b'\xf2y\xe6\xa1\xf5\xe3 \xcc\xa9\x115gm\x9c\xb6\xe4L\xa8\xa0\x8c\x0b\x884+\xcd\xb1\x14Oe\x11\xb5h'  # noqa: E501
+REWARD_TOPIC: Final = b'q\xba\xb6\\\xed.WPwZ\x06\x13\xbe\x06}\xf4\x8e\xf0l\xf9*In\xbfvc\xae\x06`\x92IT'  # noqa: E501
+DEPOSIT_EXECUTION_FOUR_BYTES = b'\xdb\x10\xc3\xb9'
+
+
+class UmamiDecoder(ArbitrumDecoderInterface):
+    def __init__(
+            self,
+            evm_inquirer: 'ArbitrumOneInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+
+    def _decode_deposit_withdraw_request(self, context: DecoderContext, event: 'EvmEvent') -> None:
+        """Decode an event in a deposit/withdraw request."""
+        is_deposit = context.tx_log.topics[0] == DEPOSIT_TOPIC
+        raw_amount = int.from_bytes(context.tx_log.data[0:32] if is_deposit else context.tx_log.data[32:64])  # noqa: E501
+        event.counterparty = CPT_UMAMI
+
+        if event.balance.amount == (amount := asset_normalized_value(
+                amount=raw_amount,
+                asset=event.asset.resolve_to_crypto_asset(),
+        )):  # deposit or withdraw event - balance will match amount in log data
+            if is_deposit:
+                event.event_type = HistoryEventType.DEPOSIT
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.notes = f'Deposit {amount} {event.asset.symbol_or_name()} into Umami'
+            else:
+                event.event_type = HistoryEventType.SPEND
+                event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                event.notes = f'Return {amount} {event.asset.symbol_or_name()} to Umami'
+
+        elif event.balance.amount == asset_normalized_value(
+                amount=context.transaction.value,
+                asset=event.asset.resolve_to_crypto_asset(),
+        ):  # Execution cost, balance will match transaction value
```

Comment:
What exactly is this? What are you checking here? Why are you checking against the transaction value itself? That's a fee right? It's also in case of withdrawals? I have definitely seen it for deposits.

Reply:
The only meaningful difference between the execution cost event and the deposit/withdraw event is the balance. Currently the asset is also different, with the execution cost in eth and the deposit in some other asset, but if a vault is ever added that uses eth, then the assets on the two events would be the same. So I think checking the event balance against values recorded elsewhere in the context is the only reliable way to differentiate between them.